### PR TITLE
Full update of the how-to

### PR DESCRIPTION
**Pull request by adjavon**
The goal of this PR is to have a linear how-to documentation that users can follow step by step to apply QuAC to their own data. Most notable changes are: 
- Reducing the amount of extra text in the introduction
- Adding navigation to make the step order clearer
- Making an example experiment to reduce the amount of code to be copied in the how-to
- Simplifying the language to a technical minimum 

Goals for review: Help find any missing parts, links, or text that makes going through the how-to difficult to follow. 
- Did you get from start to end without trouble? 
- Did you understand the steps? 
- With the paper on hand, would you be able to link the steps to parts in figure 1?
- Are the caveats (esp. wrt normalization) clear to you? 

Out-of-scope: 
- No functionality testing as of yet, that will come with the next milestone 
- No details about how or why it should work, which we leave for the paper.

### CODE REVIEW

File: docs/source/tutorials/data.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Within the `train`, `validation` and `test` directories, the sub-directories define class membership.
+The images must be organized in directories named after the class they belong to. 
+Only the first level of the directory structure is considered as the class. 
+
+Here is an example of a valid organization of the `train` directory in a case with two classes `class_A` and `class_B`.
+```{code-block} bash
+
+train/class_A/xxx.png
+train/class_A/xxy.png
+train/class_A/[...]/xxz.png
```
Comment (msschwartz21) [inline docs/source/tutorials/data.md:36]: What is `[...]` supposed to denote?

File: docs/source/tutorials/data.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    test/
+```
+
+## File organization
+Within the `train`, `validation` and `test` directories, the sub-directories define class membership.
+The images must be organized in directories named after the class they belong to. 
+Only the first level of the directory structure is considered as the class. 
+
+Here is an example of a valid organization of the `train` directory in a case with two classes `class_A` and `class_B`.
+```{code-block} bash
```
Comment (msschwartz21) [inline docs/source/tutorials/data.md:32]: Consider swapping this block for a file tree. I've used this [website](https://tree.nathanfriend.com/) to generate markdown friendly file trees in the past.

File: docs/source/tutorials/data.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+```{code-block} bash
+
+root/
+    train/
+    validation/
+    test/
+```
+
+## File organization
```
Comment (msschwartz21) [inline docs/source/tutorials/data.md:26]: Maybe add a comment about whether more than two classes is supported?

File: docs/source/tutorials/train_yaml.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    reference: "</path/to/your/source/data/val>" 
+    img_size: 128
+    batch_size: 16
+    num_workers: 12
+    mean: 0.5
+    std: 0.5
+    grayscale: true
+```
+
+- The `source` and `reference` values hold the (absolute) path your data. The data in `source` is used as the **query** image, and the data in `reference` as the **reference** image.
```
Comment (msschwartz21) [inline docs/source/tutorials/train_yaml.md:35]: Maybe add a callout with any guidance on picking source and reference classes even if it's only a recommendation to go both directions

Comment (msschwartz21) [reply docs/source/tutorials/train_yaml.md:35]: Should source and reference point to a class directory or to the split directory that contains all of the classes? If the latter, why do you need to define source and reference separately?

File: docs/source/tutorials/train_yaml.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    style_dim: 64
+    latent_dim: 16
+    num_domains: 3
+    input_dim: 1
+    final_activation: "tanh"
+```
+
+- Set `img_size` to the same value as above, in the data loading configuration. 
+- `style_dim` defines the size of the learned style space. This is the latent representation of class features present in an image, and is used to condition the conversion of images from one class to another. Small, simple datasets can have a smaller style style. 
+- `latent_dim` defines the size of the randomly sampled value from which `style` is made. It can be smaller than `style_dim`.
```
Comment (msschwartz21) [inline docs/source/tutorials/train_yaml.md:62]: Is it possible to provide any more guidance on these value? Maybe a callout with something like:
- If you see x during training/inference, that's a sign to decrease a
- If you see y during training/inference, increase b

File: docs/source/tutorials/train_yaml.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+- `validation_config` defines how we will use the classifier to validate the quality of our conversion network. 
+    - `classifier_checkpoint` points to the path where your torchscript classifier is. This should be the path you decided on [when you prepared your classifier](classifer.md)
+    - `do_nothing` is a boolean value (`true` or `false`) that defines whether to modify the output of the conversion network before you pass it to the classifier for validation. If you have correctly configured the conversion model to output the range of data that your classifier expects, then you should set this to `true`. If, however, your classifier requires a different normalization, then you should set it to `false`
+
+```{note}
+If you set `do_nothing` to `false`, you will have to set the `mean` and `std` arguments in the `validation_config`. The output of the conversion model will the be first shifted to `[0, 1]` and then we will apply `torchvision`-style normalization: `x = (x - mean) / std`.
+
+For example: if your data is in `[-1, 1]`, the following configurations are equivalent: 
+    1. `do_nothing: true`
+    2. `do_nothing: false, mean: 0.5, std: 0.5`
```
Comment (msschwartz21) [inline docs/source/tutorials/train_yaml.md:92]: List isn't rendering correctly with the current indentation
```suggestion
1. `do_nothing: true`
2. `do_nothing: false, mean: 0.5, std: 0.5`
```

File: docs/source/tutorials/train_yaml.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+- All of the arguments in `run` are defined in number of batches:
+    - `total_iters` corresponds to the total number of batches to train on. This is how you set the length of training. 
+    - Every `log_every` batches, we will log losses, some metrics, and input/output images from the `train` set to `wandb`. You should see them appear over time. 
+    - Every `save_every` batches, we will save a model checkpoint. 
+    - Every `eval_every` batches, a validation loop is run. This will go through the validation dataset, run conversion from each class to every other class, and pass the output images to the pre-trained classifier for prediction. The resulting validation metrics will also be saved to `wandb`. Note that if you have many classes and a large `validation` dataset, `validation` can be a lengthy process.  
+
+## Training
+Once you have fully edited the `config.yaml` file, you are ready to start a training run. 
+In your experiment directory, simply run `python train.py`.
```
Comment (msschwartz21) [inline docs/source/tutorials/train_yaml.md:127]:
```suggestion
In your experiment directory, simply run `python train_stargan.py`. 
```

File: docs/source/tutorials/generate.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+## Run image generation 
+
+The `generate_images.py` script will use the model that you have chosen to try to convert your images from one class to another. 
+
+It requires an additional set of parameters: 
+- `dataset`: Which of the datasets to run the translation on. By default this will be the "test" dataset, if that does not exist it will revert to the "validation" dataset.
+- `source_class`: The name of the class to take images from.
+- `target_class`: The name of the class to convert images to.
+- `checkpoint_iter`: The number of the checkpoint you chose above.
```
Comment (msschwartz21) [inline docs/source/tutorials/generate.md:26]: Just to clarify, this script is going to look for a model to load in the current working directory?

File: docs/source/tutorials/generate.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+For example, to convert the `validation` dataset from `class_A` to `class_B` using checkpoint `50000` you would run: 
+```{code-block} bash
+python generate_images.py --dataset validation --source_class class_A --target_class class_B --checkpoint_iter 50000
+```
+
+If you want to have a look at the other modifiable options, run the following: 
+```{code-block} bash
+python generate_images.py --help
+```
```
Comment (msschwartz21) [inline docs/source/tutorials/generate.md:36]: Add a comment about what the output of this script is, e.g. this script will create a new folder in the validation folder `foldername` with images converted from A to B.

File: docs/source/tutorials/attribute.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+We will get candidate regions of change by running several discriminative attribution methods using the `run_attribution.py` script. 
+It takes the following arguments, all of which are optional: 
+- `dataset`: Which of the datasets to run the translation on. By default this will be the "test" dataset, if that does not exist it will revert to the "validation" dataset.
+- `output`: Where to store the generated attributions. They will be further organized as sub-directories named after the different methods that we will try. By default, the attributions will go in the experiment root directory under `attributions`.
+- `input_fake`: Where the generated images are. You should only set these if you used a custom `output` argument in the previous step. By default, this is in the experiment root directory under `generated_images`
+
+The metadata regarding the real input images will be read from the configuration file, as before.
+
+To usefrom the `config.yaml` defaults:
```
Comment (msschwartz21) [inline docs/source/tutorials/attribute.md:16]:
```suggestion
To use the `config.yaml` defaults:
```

File: docs/source/tutorials/evaluate.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+To run using the defaults, simply run: 
+```{code-block} bash
+python run_evaluation.py
+```
+
+You can use the following command to get help with formatting your arguments.
+```{code-block} bash
+python run_evaluation.py -h
+```
```
Comment (msschwartz21) [inline docs/source/tutorials/evaluate.md:21]: I think a file tree with expected outputs would be helpful here too

File: docs/source/tutorials/visualize.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+fig, ax = plt.subplots()
+for method, report in reports.items():
+    report.plot_curve(ax=ax)
+# Add the legend
+plt.legend()
+plt.show()
+```
+
+The plots show the mean and interquartile range of QuAC curves for all examples in your testing dataset. The closer the curve gets to the upper left corner, the better.
```
Comment (msschwartz21) [inline docs/source/tutorials/visualize.md:49]: Can you include an example plot here even if it's just an illustration not on real data?

File: docs/source/tutorials/visualize.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
+# Visualizing the results
```
Comment (msschwartz21) [inline docs/source/tutorials/visualize.md:1]: I know this is probably a pain, but I found this pretty hard to follow on anything more than the most superficial level without visuals of the plots. I think it would really benefit from embedding outputs even if it's something you run once locally and just embed the plots as images.

File: docs/source/tutorials/classifier.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
+# The classifier
+
+## Training a classifier 
+
+The purpose of QuAC is to explain the decisions of a pre-trained classifier.
```
Comment (msschwartz21) [inline docs/source/tutorials/classifier.md:5]: Can you offer any guidance on minimum performance for the classifier in order for quac to work?

**Review by msschwartz21 — COMMENTED** [review]
Everything was really clear and easy to follow so your work has paid off as far as I'm concerned :tada:. For the first set of scripts, you caught all of the obvious gotchas that I could think of that aren't quac specific. 

I left comments (and direct suggestions when possible) in the files themselves and then a few higher level thoughts below. 

- To what extent do you want to cite things in the docs? The first thing that caught my eye as benefiting from a citation was the reference to StarGAN.
- A glossary could be helpful. Myst has a nice framework for it https://mystmd.org/guide/glossaries-and-terms. I can generate a list of terms to include if you want to pursue this.
- This is way overkill, but it would be very cool to make the wandb run for example data public so that the reader can follow along when you're talking about picking a StarGAN checkpoint based on the wandb logs.
- Somewhere early in the tutorial, it needs to be stated that all of the scripts necessary to run the code are available in `example_experiment` and include a link to that folder on github.
- Is there a reason why the scripts in `example_experiment` don't just live in a scripts folder? Are you expecting that the python files would get copied into each new experiment folder?

Miscellaneous suggestions
- For sanity, delete the mkdocs.yml if it's not being used
- There's a useful pre-commit hook for catching typos: https://github.com/crate-ci/typos/blob/master/docs/pre-commit.md

File: docs/source/tutorials/data.md (hunk 1 above)
Comment (adjavon) [reply docs/source/tutorials/data.md:36]: It's any arbitrary number of sub-directories, but I'll make it more explicit with a tree!

File: docs/source/tutorials/data.md (hunk 3 above)
Comment (adjavon) [reply docs/source/tutorials/data.md:26]: Decided to use 3 classes in the example, instead!

File: docs/source/tutorials/generate.md (hunk 8 above)
Comment (adjavon) [reply docs/source/tutorials/generate.md:26]: The description of where the checkpoints end up (in `train_yaml.md`) should make it clearer now where the checkpoints are expected.

File: docs/source/tutorials/train_yaml.md (hunk 4 above)
Comment (adjavon) [reply docs/source/tutorials/train_yaml.md:38]: The latter, and this was useful for previous experiments but I'll be removing this in #29 , hopefully it'll be simpler then.

File: docs/source/tutorials/visualize.md (hunk 13 above)
Comment (adjavon) [reply docs/source/tutorials/visualize.md:1]: This is good to know! But I'll be revamping this section after #44 , so I'll leave it as is for now.
